pkg/pubsub/adapter/converters/auditlog.go

```diff
@@ -0,0 +1,151 @@
+	"bytes"
+	"context"
+	errors "errors"
+	"fmt"
+	"log"
+	"reflect"
+	"regexp"
+	"strings"
+
+	cloudevents "github.com/cloudevents/sdk-go"
+	cepubsub "github.com/cloudevents/sdk-go/pkg/cloudevents/transport/pubsub"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/google/knative-gcp/pkg/apis/events/v1alpha1"
+	auditpb "google.golang.org/genproto/googleapis/cloud/audit"
+	logpb "google.golang.org/genproto/googleapis/logging/v2"
+)
+
+const (
+	AuditLogAdapterType = "google.auditlog"
+
+	logEntrySchema = "type.googleapis.com/google.logging.v2.LogEntry"
+	loggingSource  = "logging.googleapis.com"
+
+	parentResourcePattern = `^(:?projects|organizations|billingAccounts|folders)/[^/]+`
+
+	serviceNameExtension  = "servicename"
+	methodNameExtension   = "methodname"
+	resourceNameExtension = "resourcename"
+)
+
+var (
+	jsonpbUnmarshaller = jsonpb.Unmarshaler{
+		AllowUnknownFields: true,
+		AnyResolver:        resolver(resolveAnyUnknowns),
+	}
+	jsonpbMarshaler      = jsonpb.Marshaler{}
+	parentResourceRegexp *regexp.Regexp
+)
+
+func init() {
+	var err error
+	if parentResourceRegexp, err = regexp.Compile(parentResourcePattern); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// Resolver function type that can be used to resolve Any fields in a jsonpb.Unmarshaler.
+type resolver func(turl string) (proto.Message, error)
+
+func (r resolver) Resolve(turl string) (proto.Message, error) {
+	return r(turl)
+}
+
+type UnknownMsg empty.Empty
+
+func (m *UnknownMsg) ProtoMessage() {
+	(*empty.Empty)(m).ProtoMessage()
+}
+
+func (m *UnknownMsg) Reset() {
+	(*empty.Empty)(m).Reset()
+}
+
+func (m *UnknownMsg) String() string {
+	return "Unknown message"
+}
+
+// Resolver type which resolves unknown message types to empty.Empty.
+func resolveAnyUnknowns(typeURL string) (proto.Message, error) {
+	// Only the part of typeUrl after the last slash is relevant.
+	mname := typeURL
+	if slash := strings.LastIndex(mname, "/"); slash >= 0 {
+		mname = mname[slash+1:]
+	}
+	mt := proto.MessageType(mname)
+	if mt == nil {
+		return (*UnknownMsg)(&empty.Empty{}), nil
+	}
+	return reflect.New(mt.Elem()).Interface().(proto.Message), nil
+}
+
+func convertAuditLog(ctx context.Context, msg *cepubsub.Message, sendMode ModeType) (*cloudevents.Event, error) {
+	if msg == nil {
+		return nil, fmt.Errorf("nil pubsub message")
+	}
+	entry := logpb.LogEntry{}
+	if err := jsonpbUnmarshaller.Unmarshal(bytes.NewReader(msg.Data), &entry); err != nil {
+		return nil, fmt.Errorf("failed to decode LogEntry: %w", err)
+	}
+
+	parentResource := parentResourceRegexp.FindString(entry.LogName)
+	if parentResource == "" {
+		return nil, fmt.Errorf("invalid LogName: %q", entry.LogName)
+	}
+
+	// Make a new event and convert the message payload.
+	event := cloudevents.NewEvent(cloudevents.VersionV1)
+	event.SetID(entry.InsertId + entry.LogName + ptypes.TimestampString(entry.Timestamp))
```

Comment:
Would it be helpful to add separator in between?

Reply:
I don't think id is meant to be human readable so I think this is unnecessary.